_posts/2021-01-04-unit-tests-and-concurrency.md

```diff
@@ -0,0 +1,135 @@
+However the production code usually is not single threaded. IO operations are done on the IO thread pool, views are updated on the main thread and everything else happens on the computation pool. By using different schedulers in the tests and using a different strategy (single threaded) we make those unit tests useless in catching concurrency issues. 
+
+## A real world scenario
+
+I was working on a side project. The screen consists of a `RecyclerView` displaying a list of elements. To get the elements I need to perform two different API calls. The first API call returns a list with N elements, then for each item in the list I need to perform the second call. After combining the data I send it to the UI for displaying.
+Using RxJava this looks like:
+
+```kotlin
+// Emits Loading then Content or Problem
+private fun requestData(): Observable<ViewState> =
+    service.firstApiCall()
+        .observeOn(Schedulers.computation())
+        .map { it.message }
+        .flatMap(this::secondApiCall)
+        .map<ViewState> { ViewState.Content(it) }
+        .startWith(Single.just(ViewState.Loading))
+        .onErrorReturn { ViewState.Problem }
+        .toObservable()
+
+// Concurrently executes secondApiCall for each element in list. 
+// Transforms the result to ViewEntity, combines everything in a list
+private fun secondApiCall(list: List<String>): Single<List<ViewEntity>> =
+    Observable.fromIterable(list)
+        .concatMapEager { item ->
+            service.secondEndpoint(item)
+                .observeOn(Schedulers.computation())
+                .map { it.message }
+                .map { ViewEntity(item, it.toHttpUrl()) }
+                .onErrorReturn { DogViewEntity(item, null) }
+                .toObservable()
+        }
+        .toList()
+```
+
+Since the list is 15-20 elements, I am using `concatMapEager` to execute the API calls for the second endpoint concurrently while maintaining the order of the elements. This worked great and I wrote my unit tests using the real schedulers. RxJava has amazing testing support that enables this. Thanks [Simon Vergauwen][simon] for the tips on how to do that.
+
+```kotlin
+@Test
+fun `execute - first call succeeds, one failed call to #2 - one full item, one item with default value`() {
+    val items = listOf("Item 1" to "https://item1.jpg", "Item 2" to "invalid url")
+    val service = successService(items)
+
+    val viewModel = ViewModel(service)
+
+    val expected = listOf(
+        ViewEntity("Item 1", "https://item1.jpg".toHttpUrl()),
+        ViewEntity("Item 2", null)
+    )
+    viewModel.state
+        .test()
+        .also { viewModel.execute() }
```

Comment:
yeah, I need to `subscribe` before I act (otherwise I lose the default value) and act before I can assert.....

another alternative is to use `ReplaySubject` in tests